openedx/core/djangoapps/course_groups/cohorts.py

```diff
@@ -235,47 +211,72 @@ def get_cohort(user, course_key, assign=True):
         if not assign:
             return None
 
-    choices = course.auto_cohort_groups
-    if len(choices) > 0:
-        # Randomly choose one of the auto_cohort_groups, creating it if needed.
-        group_name = local_random().choice(choices)
+    cohorts = get_course_cohorts(course, assignment_type=CourseCohort.RANDOM)
+    if cohorts:
+        cohort = local_random().choice(cohorts)
     else:
-        # Use the "default cohort".
-        group_name = DEFAULT_COHORT_NAME
+        cohort, __ = CourseUserGroup.create(name=DEFAULT_COHORT_NAME, course_id=course_key)
+        CourseCohort.create(course_user_group=cohort, assignment_type=CourseCohort.RANDOM)
 
-    group, __ = CourseUserGroup.objects.get_or_create(
-        course_id=course_key,
-        group_type=CourseUserGroup.COHORT,
-        name=group_name
+    user.course_groups.add(cohort)
+
+    return cohort
+
+
+def migrate_cohort_settings(course):
+    """
+    Migrate all the cohort settings associated with this course from modulestore to mysql.
+    After that we will never touch modulestore for any cohort related settings.
+    """
+    course_id = course.location.course_key
+    cohort_settings, created = CourseCohortsSettings.objects.get_or_create(
+        course_id=course_id,
+        defaults={
+            'is_cohorted': course.is_cohorted,
+            'cohorted_discussions': json.dumps(list(course.cohorted_discussions)),
+            'always_cohort_inline_discussions': course.always_cohort_inline_discussions
+        }
     )
-    user.course_groups.add(group)
-    return group
 
+    # Add the new and update the existing cohorts
+    if created:
+        # Update the manual cohorts already present in CourseUserGroup
+        manual_cohorts = CourseUserGroup.objects.filter(
+            course_id=course_id,
+            group_type=CourseUserGroup.COHORT
+        ).exclude(name__in=course.auto_cohort_groups)
+        for cohort in manual_cohorts:
+            CourseCohort.create(course_user_group=cohort)
+
+        for group_name in course.auto_cohort_groups:
+            cohort, __ = CourseUserGroup.create(name=group_name, course_id=course_id)
+            CourseCohort.create(course_user_group=cohort, assignment_type=CourseCohort.RANDOM)
+
```

Comment:
In the first `for loop`, we are just creating cohort assignment types in CourseCohort for cohorts present in CourseUserGroup. In this case i don't think there will be anything in CourseCohort so i think there will be no issue.

In the second `for loop`, we are creating/migrating auto cohorts to CourseUserGroup and CourseCohort. Lets say if some auto cohort already exists in CourseUserGroup then we will not create it.

Do you see a scenario in which this can cause issue?